Python interface design documents (#8)
pace-neutrons/pace-developers: python_interface/design/01_pace_python_high_level_discussion.md

# Matlab and Python: A prototype design for PACE
**2020-03-23**

# Scope

This document summarises, at a high level, methods for running interdependent Matlab and Python code alongside each other, with and without a Matlab license.
Implementation details are presented in a [separate document](02_pace_python_implementation_discussion.md).

# Motivation

> PACE will be a suit of several interconnected programs for inelastic neutron data analysis, which are written both in Matlab and in Python.

Reviewer: ```suggestion
PACE will be a suite of several interconnected programs for inelastic neutron scattering data analysis, which are written both in Matlab and in Python.
```

These techniques rely on the client library being loaded into the same memory as the host, and use C `mex` or python module code to call the Matlab or Python functions directly.
An alternative which would use only pure Matlab or Python code (without needing any compilation) is to use a [reverse-communication interface](#reverse-communication-interface).
The implications of these proposed implementations on parallelisation is discussed in the [final section](#parallelisation).
We end with a [summary and discussion points](#summary).

# <a name="python-interface"></a> Python User Interface

As detailed in [this section](02_pace_python_implementation_discussion.md#compiled_matlab), Matlab code may be "compiled" using the **Matlab Compiler** and freely distributed to users without a Matlab license.
In order to run this "compiled" code, the user is required to install the free/gratis **Matlab Compiler Runtime** package, and this **MCR** package must match the version of Matlab used to compile the Horace library.

> Thus, if we always target the latest version of Matlab, we should require users to re-install (upgrade) the **MCR** twice a year, which may not be desirable.

Reviewer 2: I'm not sure there is any benefit in following the latest MATLAB releases. The compiled route is a self-contained system so we can upgrade when it is beneficial to do so (new features, bug fixes in the MCR, ...).

Author: ```suggestion
Thus, it would make sense to target a specific Matlab release for stability until new features make upgrading desirable.
```

The compiled Matlab library is a platform-independent Component Technology File (`ctf`) which contains compressed and encrypted versions of the `m` and `mex`-files used in the project.

> The **MCR** is a platform-dependent installation (approximately 1GB download) which contains the base Matlab distribution and libraries but is set up to be able to run only code from `ctf`s.

Reviewer 2: Would it be helpful to add a link to the MCR download site: https://uk.mathworks.com/products/compiler/matlab-runtime.html

Author: ```suggestion
The **MCR** is a platform-dependent installation ([approximately 1GB download](https://uk.mathworks.com/products/compiler/matlab-runtime.html)) which contains the base Matlab distribution and libraries but is set up to be able to run only code from `ctf`s.
```

Data types which are not basic scalar types or arrays of such, including Python functions and objects are not allowed to be passed to Matlab, and will raise an error if tried.
This has implications for calling Python functions from Matlab using the Python interface, as discussed in the next section.

Since array data within a `matlab.object` is *not* copied when exported from Matlab to Python, we may use it to wrap large data arrays, to avoid the data copying discussed in the previous section.
However, the reverse direction is more problematic since the Matlab-Python library does not allow arbitrary Python objects to be exported to Matlab.
Thus large array exports from Python to Matlab will need either to be copied using the default export mechanism, or we need to implement the shared-data [export mechanism described in the details](02_pace_python_implementation_discussion.md#mex_wrapping).
This will definitely be needed if large parts of PACE come to be written in Python, and only small parts remain in Matlab and will thus need to communicate much more data with Python.

> As a bonus feature, the share-copies mechanism will allow users fast access to the `pix` array of an `sqw` object for example.

Reviewer 2: The plan for Horace was to change the primary interaction with the pixel array from direct array access to column-based getters and setter, i.e. rather than accessing `sqw.data.pix[6]` the user should call `sqw.data.getSignal()` or `sqw.data.getData('signal')`.
Does that design decision have implications on here?

Author: I'm not sure - in principle no. I think the Python interface will respect `Access = private` attributes on Matlab objects so this will not be visible to Python users either, just the setter/getter; but I need to check. The getter/setter will be accessed in Python just the same as in Matlab.

## Example syntax

Finally, whilst the [prototype implementation](02_pace_python_implementation_discussion.md#user_interface) has a basic Python wrapper for inferring function calls and automatic type conversions which allows for syntax such as:

```python
from pyHorace import Matlab
m = Matlab()
proj = m.projaxes([ -0.5, 1, 0], [0, 0, 1], 'type', 'rrr')
w1 = m.cut_sqw('ei30_10K.sqw', proj, [0.1, 0.02, 0.5], [1.5, 2.5], [0.4, 0.5], [3, 0.5, 20])
w2 = w1.cut([0.1, 0.5], [3, 0.5, 20])   # or m.cut(w1, [0.1, 0.5], [3, 0.5, 20])
w2.plot()                               # or m.plot(w2)
```

Note that it is not possible to import all Matlab functions or classes into the Python namespace directly so syntax which omits the `m.` will only be possible by explicitly specifying individual Python wrappers to equivalent Matlab functions.

# <a name="interpreter-calls"></a> Calling user defined functions

> Depending on the outcome of the design review for the optimiser, described [here](/optimisation/design/Model_Optimisation_Design.md), the optimiser may be written in Matlab or in Python (or perhaps C++).

Reviewer 2: You look at the MATLAB and Python optimiser use-cases. How would a pure C++ optimiser be used?

Author: It would need a wrapper each for the Python (probably using PyBind11) and Matlab interfaces (using either mex or if we decide to use Matlab R2019a or newer, the `clibgen` functions to export a set of headers so Matlab can call the C++ library directly - see https://blogs.mathworks.com/developer/2019/07/11/cpp-interface/ ).

For data exchange, on the Matlab <-> C++ front, we would probably have to write mex functions similar to what would be needed if the optimiser was written in Python (albeit probably simpler than if it was in Python) to ensure that the data was not needlessly copied. I don't know how `clibgen` works in this respects though. On the Python <-> C++ front, PyBind11 does all the heavy lifting so it would be even simpler.
